src/main/java/org/candlepin/resource/util/ConsumerCloudDataBuilder.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2009 - 2025 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.resource.util;
+
+import org.candlepin.dto.api.server.v1.ConsumerDTO;
+import org.candlepin.exceptions.BadRequestException;
+import org.candlepin.model.Consumer;
+import org.candlepin.model.ConsumerCloudData;
+
+import org.xnap.commons.i18n.I18n;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+/**
+ * Builder for constructing {@link ConsumerCloudData} from consumer facts.
+ * <p>
+ * This class selects the appropriate {@link CloudProviderFactParser} based on the provided facts and
+ * builds a {@link ConsumerCloudData} object containing cloud-specific information.
+ * </p>
+ */
+public class ConsumerCloudDataBuilder {
```

Comment:
We can use a `Singleton` annotation on this class

Reply:
Done